src/coreclr/jit/fgopt.cpp

```diff
@@ -4522,6 +4537,314 @@ bool Compiler::fgReorderBlocks(bool useProfile)
+// fgDoReversePostOrderLayout: Reorder blocks using a greedy RPO traversal.
+//
+void Compiler::fgDoReversePostOrderLayout()
+{
+#ifdef DEBUG
+    if (verbose)
+    {
+        printf("*************** In fgDoReversePostOrderLayout()\n");
+
+        printf("\nInitial BasicBlocks");
+        fgDispBasicBlocks(verboseTrees);
+        printf("\n");
+    }
+#endif // DEBUG
+
+    // Compute DFS of main function body's blocks, using profile data to determine the order successors are visited in
+    //
+    FlowGraphDfsTree* const dfsTree = fgComputeDfs</* useProfile */ true>();
+
+    // Fast path: We don't have any EH regions, so just reorder the blocks
+    //
+    if (compHndBBtabCount == 0)
+    {
+        for (unsigned i = dfsTree->GetPostOrderCount() - 1; i != 0; i--)
+        {
+            BasicBlock* const block       = dfsTree->GetPostOrder(i);
+            BasicBlock* const blockToMove = dfsTree->GetPostOrder(i - 1);
+            fgUnlinkBlock(blockToMove);
+            fgInsertBBafter(block, blockToMove);
+        }
+
+        return;
+    }
+
+    // The RPO will break up call-finally pairs, so save them before re-ordering
+    //
+    BlockToBlockMap callFinallyPairs(getAllocator());
+
+    for (BasicBlock* const block : Blocks())
+    {
+        if (block->isBBCallFinallyPair())
+        {
+            callFinallyPairs.Set(block, block->Next());
+        }
+    }
+
+    // Reorder blocks
+    //
+    for (unsigned i = dfsTree->GetPostOrderCount() - 1; i != 0; i--)
+    {
+        BasicBlock* const block       = dfsTree->GetPostOrder(i);
+        BasicBlock* const blockToMove = dfsTree->GetPostOrder(i - 1);
+
+        // Only reorder blocks within the same EH region -- we don't want to make them non-contiguous
+        //
+        if (BasicBlock::sameEHRegion(block, blockToMove))
+        {
+            // Don't reorder EH regions with filter handlers -- we want the filter to come first
+            //
+            if (block->hasHndIndex() && ehGetDsc(block->getHndIndex())->HasFilter())
+            {
+                continue;
+            }
+
+            fgUnlinkBlock(blockToMove);
+            fgInsertBBafter(block, blockToMove);
+        }
+    }
+
+    // Fix up call-finally pairs
+    //
+    for (BlockToBlockMap::Node* const iter : BlockToBlockMap::KeyValueIteration(&callFinallyPairs))
+    {
+        BasicBlock* const callFinally    = iter->GetKey();
+        BasicBlock* const callFinallyRet = iter->GetValue();
+        fgUnlinkBlock(callFinallyRet);
+        fgInsertBBafter(callFinally, callFinallyRet);
+    }
+
+    // The RPO won't change the entry blocks of any EH regions, but reordering can change the last block in a region
+    // (for example, by pushing throw blocks unreachable via normal flow to the end of the region).
+    // First, determine the new EH region ends.
+    //
+    BasicBlock** const tryRegionEnds       = new (this, CMK_Generic) BasicBlock* [compHndBBtabCount] {};
```

Comment:
Consider using `ArrayStack<struct>` for this data (where `struct` contains your 3 data items). It will usually be able to avoid heap allocation.